exchanges/stream/websocket.go

```diff
@@ -786,163 +776,134 @@ func (w *Websocket) GetName() string {
 
 // GetChannelDifference finds the difference between the subscribed channels
 // and the new subscription list when pairs are disabled or enabled.
-func (w *Websocket) GetChannelDifference(genSubs []subscription.Subscription) (sub, unsub []subscription.Subscription) {
-	w.subscriptionMutex.RLock()
-	unsubMap := make(map[any]subscription.Subscription, len(w.subscriptions))
-	for k, c := range w.subscriptions {
-		unsubMap[k] = *c
-	}
-	w.subscriptionMutex.RUnlock()
-
-	for i := range genSubs {
-		key := genSubs[i].EnsureKeyed()
-		if _, ok := unsubMap[key]; ok {
-			delete(unsubMap, key) // If it's in both then we remove it from the unsubscribe list
-		} else {
-			sub = append(sub, genSubs[i]) // If it's in genSubs but not existing subs we want to subscribe
-		}
-	}
-
-	for x := range unsubMap {
-		unsub = append(unsub, unsubMap[x])
+func (w *Websocket) GetChannelDifference(newSubs subscription.List) (sub, unsub subscription.List) {
+	if w.subscriptions == nil {
+		w.subscriptions = subscription.NewStore()
 	}
-
-	return
+	return w.subscriptions.Diff(newSubs)
 }
 
-// UnsubscribeChannels unsubscribes from a websocket channel
-func (w *Websocket) UnsubscribeChannels(channels []subscription.Subscription) error {
+// UnsubscribeChannels unsubscribes from a list of websocket channel
+func (w *Websocket) UnsubscribeChannels(channels subscription.List) error {
 	if len(channels) == 0 {
-		return fmt.Errorf("%s websocket: %w", w.exchangeName, errNoSubscriptionsSupplied)
+		return errNoSubscriptionsSupplied
 	}
-	w.subscriptionMutex.RLock()
-
-	for i := range channels {
-		key := channels[i].EnsureKeyed()
-		if _, ok := w.subscriptions[key]; !ok {
-			w.subscriptionMutex.RUnlock()
-			return fmt.Errorf("%s websocket: %w: %+v", w.exchangeName, ErrSubscriptionNotFound, channels[i])
+	if w.subscriptions == nil {
+		return common.ErrNilPointer
```

Comment:
suggestion: `fmt.Errorf("%w for %T", err, w.subscriptions)` not likely its going to happen but if it does then you know what's going down, can be done to all your common.ErrNilPointer.

Reply:
As discussed, but I'll try to be briefer:
The boundary of this function is to remove channels and not panic.
It should not be asserting "correctness" of the state of w and rejecting no subscriptions, because there's no reason for it to.
It shouldn't need to check the length and error if 0, because that's not really an error.
It's a subjective view that Unsubscribing from an empty list is an error, when really it's not,
it's a convenience.
Removing the len() check means neither the caller or us need to check len and handle error, and the unsubscribe will return okay. Allowing a user to call Unsub without checking what it's unsubbing from isn't invalid.

Overall:
Happy to protect against nil panics, but that might take the form of lazy init as often as returning an error.

Fixed: 1040f37

Reply:
Done. I don't think the type is informative enough, if we're going to go to the trouble of providing any context.

I'm really not comfortable with this, but I want this merged so I'll make a note to come back to the topic better once we've documented guidelines.